---
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r, load-libraries}
library(ggalluvial)
library(patchwork)
library(tidyverse)
```


Sankey plot to show categories of data
```{r}
## Load xlxs of relationship between primary sources, OF, and inference
sankey.data <- readxl::read_excel("/drive-pool/data/peter_data/repo/genetic_algorithm_gwas_sc/gwas_gp_manu_data/sankey.data.catalog_revised.xlsx")
## Add freq because sankey plots needs to know quantity
sankey.data$freq <- 1
## Convert to of names to boolean style function
map.of <- data.frame(axis2 = c("Marker Gene", "MAGMA", "Cancer Gene", "PPI", "Marker PPI", "Inter-PPI",
                            "Intra-PPI", "LPI", "Promoter", "Marker ATAC", "Common ATAC"),
                     new.of = c("isMarkerGene", "isMAGMAgene", "isCancerGene", "isPPI", "isMarkerPPI", "isInterPPI",
                                "isIntraPPI", "isLPI", "isPromoter", "isMarkerATAC", "isCommonATAC"))

## Update snRNA/ATAC to have seq
sankey.data[which(sankey.data$primary == "snRNA"), "primary"] <- "snRNA-seq"
sankey.data[which(sankey.data$primary == "snATAC"), "primary"] <- "snATAC-seq"

## Set column order
sankey.data <- sankey.data %>% left_join(x = ., y = map.of, by = "axis2")
sankey.data$new.of <- factor(x = sankey.data$new.of, levels = c("isMAGMAgene", "isCancerGene", "isPPI", "isLPI", "isPromoter",
                                                                "isCommonATAC", "isMarkerATAC", "isMarkerGene", "isMarkerPPI",
                                                                "isIntraPPI", "isInterPPI"))
sankey.data$primary <- factor(x = sankey.data$primary, levels = c("MAGMA", "Cosmic CGC", "StringDB", "LncBook", "GRCh38",
                                                                  "snRNA-seq", "snATAC-seq", "CellTalkDB"))
sankey.data$axis3 <- factor(x = sankey.data$axis3, levels = c("Gene", "Cell"))

## Load kelly color for annotation
load("/drive-pool/data/peter_data/repo/genetic_algorithm_gwas_sc/gwas_gp_manu_data/kelly.colours.rda")

p1 <- ggplot(sankey.data,
       aes(y = freq,
           axis1 = primary, axis2 = new.of, axis3 = axis3)) +
  geom_alluvium(aes(fill = primary)) + ## Annotate the OF number
  geom_stratum(alpha = 0.5, width = c(rep(0.325,times=8), rep(.45,times=11), rep(0.325,times=2))) + ## Custom width size
  geom_text(stat = "stratum", aes(label = after_stat(stratum)), size = 3, family = "Arial") +
  ylab("") +
  theme_minimal() +
  scale_fill_manual(values = kelly.colours[c(7,8,5,6,4,3,10,12)]) +
  scale_x_continuous(breaks = 1:3, labels = c("Primary\nData Sources", "Objective\nFunction", "Inference")) +
  theme(panel.grid = element_blank(),
        axis.text.y = element_blank(),
        axis.text.x = element_text(size = 8),
        axis.title = element_blank(),
        text = element_text(family = "Arial"),
        legend.position = "none")

layout <- "
AAAAAAAA
AAAAAAAA
AAAAAAAA
AAAAAAAA
BBBBBBBB
BBBBBBBB
BBBBBBBB
BBBBBBBB
"


dev.off()
combined.plot <- p1 + plot_spacer() + plot_layout(design = layout) &
  theme(
        plot.tag = element_text(face = 'bold', size = 10),
        axis.text = element_text(size = 8),
        axis.title = element_text(size = 8),
        legend.text = element_text(size = 8),
        legend.title = element_text(size = 8),
        text = element_text(family = "Arial")) 
ggsave("/drive-pool/data/peter_data/repo/genetic_algorithm_gwas_sc/figures/figure_1.png", width = 170, height = 170,
       units = "mm", dpi = 600)
```
